File: hello_ollama/agent.py
import logging
from google.adk.models.lite_llm import LiteLlm
from google.adk.agents import Agent

from hello_ollama.periodic_table import periodic_map


def get_element_info(symbol: str) -> dict:
    """Returns information about a chemical element by its Chinese name."""
    logging.info(f"查询元素信息: {symbol}")
    if symbol in periodic_map():
        element = periodic_map()[symbol]
    else:
        logging.warning(f"元素符号或中文名 '{symbol}' 未找到。")
        return {"status": "error", "error_message": f"元素符号或中文名 '{symbol}' 未找到。"}

    chinese_name = element['chinese_name']
    element_name = element['name']
    atomic_number = element['atomic_number']
    atomic_weight_value = element['atomic_weight']
    report = (
        f"{chinese_name}（{element_name}），原子序数：{atomic_number}，原子量：{atomic_weight_value}"
    )
    logging.info(f"查询成功: {report}")
    return {"status": "success", "report": report}


root_agent = Agent(
    # ollama
    # model=LiteLlm(model="ollama_chat/qwen2.5"),
    # google
    model="gemini-2.0-flash",
    name="ollama_agent",
    description=(
        "Agent to answer questions about chemical element information."
    ),
    instruction=(
        "You are a helpful agent."
    ),
    tools=[get_element_info]
)

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s %(message)s')

File: hello-adk-ts/agent.ts
import { FunctionTool, LlmAgent } from '@google/adk';
import { z } from 'zod';
import { getElement } from './periodic-table';
import { LiteLlm } from './lite-llm';

/* Periodic Table Element Info Tool */
const getElementInfoTool = new FunctionTool({
  name: 'get_element_info',
  description: 'Get information about a chemical element by symbol or Chinese name.',
  parameters: z.object({
    symbol: z
      .string()
      .describe('Element symbol (e.g., "H", "O") or Chinese name (e.g., "氢", "氧")'),
  }),
  execute: ({ symbol }: { symbol: string }) => {
    console.log(`查询元素信息: ${symbol}`);
    
    const element = getElement(symbol);
    
    if (!element) {
      const errorMsg = `元素符号或中文名称 '${symbol}' 未找到。`;
      console.warn(errorMsg);
      return {
        status: 'error',
        error_message: errorMsg,
      };
    }

    const report = `${element.chineseName}（${element.name}），原子序数：${element.atomicNumber}，原子量：${element.atomicWeight}`;
    console.log(`查询成功: ${report}`);
    
    return {
      status: 'success',
      report: report,
    };
  },
});

// Use Ollama Qwen2.5 model via LiteLlm
const adkModel = new LiteLlm('ollama_chat/qwen2.5');

export const rootAgent = new LlmAgent({
  name: 'ollama_agent',
  model: adkModel,
  description: 'Agent to answer questions about chemical element information.',
  instruction: `You are a helpful agent that provides information about chemical elements.
Use the 'get_element_info' tool to look up element information by symbol or Chinese name.
- Call this tool ONLY ONCE per element.
- Do NOT call the tool multiple times for the same element.
- Use the information returned by the tool to answer the user's question.
- If you already have the information about an element from a previous tool call, do NOT call the tool again.
- Only call the tool if you need additional information that you don't already have.
Always use the tool to get accurate information.`,
  tools: [getElementInfoTool],
});
